coderd/workspaceagents.go

```diff
@@ -676,6 +695,94 @@ func (api *API) workspaceAgentReportStats(rw http.ResponseWriter, r *http.Reques
 	}
 }
 
+func (api *API) postWorkspaceAppHealth(rw http.ResponseWriter, r *http.Request) {
+	workspaceAgent := httpmw.WorkspaceAgent(r)
+	var req codersdk.PostWorkspaceAppHealthsRequest
+	if !httpapi.Read(r.Context(), rw, r, &req) {
+		return
+	}
+
+	if req.Healths == nil || len(req.Healths) == 0 {
+		httpapi.Write(r.Context(), rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Health field is empty",
+		})
+		return
+	}
+
+	apps, err := api.Database.GetWorkspaceAppsByAgentID(r.Context(), workspaceAgent.ID)
+	if err != nil {
+		httpapi.Write(r.Context(), rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Error getting agent apps",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	var newApps []database.WorkspaceApp
+	for name, newHealth := range req.Healths {
+		old := func() *database.WorkspaceApp {
+			for _, app := range apps {
+				if app.Name == name {
+					return &app
+				}
+			}
+
+			return nil
+		}()
+		if old == nil {
+			httpapi.Write(r.Context(), rw, http.StatusNotFound, codersdk.Response{
+				Message: "Error setting workspace app health",
+				Detail:  xerrors.Errorf("workspace app name %s not found", name).Error(),
+			})
+			return
+		}
+
+		if old.HealthcheckUrl == "" {
+			httpapi.Write(r.Context(), rw, http.StatusNotFound, codersdk.Response{
+				Message: "Error setting workspace app health",
+				Detail:  xerrors.Errorf("health checking is disabled for workspace app %s", name).Error(),
+			})
+			return
+		}
+
+		switch newHealth {
+		case codersdk.WorkspaceAppHealthInitializing:
+		case codersdk.WorkspaceAppHealthHealthy:
+		case codersdk.WorkspaceAppHealthUnhealthy:
+		default:
+			httpapi.Write(r.Context(), rw, http.StatusBadRequest, codersdk.Response{
+				Message: "Error setting workspace app health",
+				Detail:  xerrors.Errorf("workspace app health %s is not a valid value", newHealth).Error(),
+			})
+			return
+		}
+
+		// don't save if the value hasn't changed
+		if old.Health == database.WorkspaceAppHealth(newHealth) {
+			continue
+		}
+		old.Health = database.WorkspaceAppHealth(newHealth)
+
+		newApps = append(newApps, *old)
+	}
+
+	for _, app := range newApps {
+		err = api.Database.UpdateWorkspaceAppHealthByID(r.Context(), database.UpdateWorkspaceAppHealthByIDParams{
```

Comment:
We could update these in one bulk request instead, similar to how we fetch by multiple IDs. That way, we wouldn't need to do any comparison logic I believe, and it'd be less DB requests.

Reply:
I wasn't sure how to do bulk updates in sqlc so I settled for this. I'm down to improve this in a separate PR if we can.